quickjs/src/main/jni/quickjs-jni.cpp

```diff
@@ -0,0 +1,329 @@
+#include <jni.h>
+#include <new>
+#include <string>
+#include <vector>
+#include "quickjs/quickjs.h"
+
+class JsMethodProxy {
+public:
+  JsMethodProxy(const char *name, jmethodID methodId)
+      : name(name), methodId(methodId) {
+  }
+
+  std::string name;
+  jmethodID methodId;
+
+  JSValueConst (**argLoaders)(JNIEnv *env, JSContext *jsContext, jvalue value);
+};
+
+class JsObjectProxy {
+public:
+  JsObjectProxy(const char *name)
+      : name(name) {
+  }
+
+  std::string name;
+  std::vector<JsMethodProxy> methods;
+};
+
+
+static jmethodID booleanValueOf;
+static jmethodID integerValueOf;
+static jmethodID doubleValueOf;
+static jmethodID quickJsExceptionConstructor;
+
+class Context {
+public:
+  Context(JNIEnv *env)
+      : env(env), jsRuntime(JS_NewRuntime()), jsContext(JS_NewContext(jsRuntime)),
+        booleanClass(static_cast<jclass>(env->NewGlobalRef(env->FindClass("java/lang/Boolean")))),
+        integerClass(static_cast<jclass>(env->NewGlobalRef(env->FindClass("java/lang/Integer")))),
+        doubleClass(static_cast<jclass>(env->NewGlobalRef(env->FindClass("java/lang/Double")))),
+        quickJsExecptionClass(static_cast<jclass>(env->NewGlobalRef(
+            env->FindClass("com/squareup/quickjs/QuickJsException")))) {
+  }
+
+  ~Context() {
+    env->DeleteGlobalRef(quickJsExecptionClass);
+    env->DeleteGlobalRef(doubleClass);
+    env->DeleteGlobalRef(integerClass);
+    env->DeleteGlobalRef(booleanClass);
+    JS_FreeContext(jsContext);
+    JS_FreeRuntime(jsRuntime);
+  }
+
+  JNIEnv *env;
+  JSRuntime *jsRuntime;
+  JSContext *jsContext;
+  jclass booleanClass;
+  jclass integerClass;
+  jclass doubleClass;
+  jclass quickJsExecptionClass;
+  std::vector<JsObjectProxy> objectProxies;
+};
+
+
+extern "C" JNIEXPORT jlong JNICALL
+Java_com_squareup_quickjs_QuickJs_createContext(JNIEnv *env, jclass type) {
+  Context *c = new(std::nothrow) Context(env);
+  if (!c) {
+    return 0;
+  }
+  booleanValueOf = env->GetStaticMethodID(c->booleanClass, "valueOf",
```

Comment:
I don't understand why this is stored in a static field whereas the class reference is stored in an instance field of the context. Shouldn't they be kept either both static or both instance?

Reply:
The method ID is managed by the VM - there's no method to explicitly release it like there is for the global class references, which is why I left them out of the context structure.  After your comment and looking at it some more though, I agree it feels weird, so I'll move them into the Context class.